src/serial.rs

```diff
@@ -251,21 +237,21 @@ macro_rules! hal {
                     apb.rstr().modify(|_, w| w.$usartXrst().set_bit());
                     apb.rstr().modify(|_, w| w.$usartXrst().clear_bit());
 
-                    // disable hardware flow control
-                    // TODO enable DMA
-                    // usart.cr3.write(|w| w.rtse().clear_bit().ctse().clear_bit());
-
                     let brr = clocks.$pclkX().0 / baud_rate.0;
                     assert!(brr >= 16, "impossible baud rate");
                     // NOTE(write): uses all bits of this register.
                     usart.brr.write(|w| unsafe { w.bits(brr) });
 
-                    // UE: enable USART
-                    // RE: enable receiver
-                    // TE: enable transceiver
-                    usart
-                        .cr1
-                        .modify(|_, w| w.ue().set_bit().re().set_bit().te().set_bit());
+                    usart.cr3.modify(|_, w| {
+                        w.dmar().enabled();  // enable DMA for reception
+                        w.dmat().enabled()   // enable DMA for transmission
+                    });
```

Comment:
Hmm, maybe we should only enable the DMA when actually using it ?

Reply:
Here I just went with what the stm32l0xx-hal is doing: https://github.com/stm32-rs/stm32l0xx-hal/blob/master/src/serial.rs#L353

I don't know what their reasoning was, but for me it boils down to:
- Setting these bits does not start any additional peripherals, so it *should* not lead to more resource usage, so we should be fine to always do it. But to be honest I don't know what setting the bits really does on a hardware level.
- We could set these bits inside `read_exact` and `write_all` instead, but that would mean unnecessary register modifies when those methods are called multiple times.

But it probably doesn't matter much in either case. I'd not be opposed to moving these DMA enables if you think it would be better that way.

Reply:
I think it starts sending DMA requests, not sure what implications this has on a disabled channel, specifically, if a old request would stay there until the DMA is enabled or if it would be instantly dropped.

Reply:
Okay, "not sure what would happen" is a good reason to err on the safe side. I'll move those enables.

Reply:
Implementing this has been a bit more work than I expected, because if we only enable DMA when we actually start a transfer, we should also disable it again when the transfer is over. So I repurposed the `Target` trait to provide `enable_dma` and `disable_dma` methods that are now used as callbacks when a `Transfer` starts and stops. The mapping from targets to DMA channels is split into a separate trait called `OnChannel` now.